File: dsa_leetcode_beginner/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> charIndexMap = new HashMap<>();
        int maxLength = 0, left = 0;

        for (int right = 0; right < s.length(); right++) {
            char currentChar = s.charAt(right);
            if (charIndexMap.containsKey(currentChar) && charIndexMap.get(currentChar) >= left) {
                left = charIndexMap.get(currentChar) + 1;
            }
            charIndexMap.put(currentChar, right);
            maxLength = Math.max(maxLength, right - left + 1);
        }

        return maxLength;
    }
}

File: dsa_leetcode_beginner/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
function lengthOfLongestSubstring(s) {
    let charIndexMap = new Map();
    let maxLength = 0;
    let left = 0;

    for (let right = 0; right < s.length; right++) {
        if (charIndexMap.has(s[right]) && charIndexMap.get(s[right]) >= left) {
            left = charIndexMap.get(s[right]) + 1;
        }
        charIndexMap.set(s[right], right);
        maxLength = Math.max(maxLength, right - left + 1);
    }

    return maxLength;
}
